src/jose/jws/general/verifier.ts

```diff
@@ -0,0 +1,112 @@
+import type { GeneralJws, Signature } from './types';
+import type { JwkPublic, VerifyFn } from '../../types';
+import type { VerificationMethod } from '../../../did/did-resolver';
+
+import { base64url } from 'multiformats/bases/base64';
+import { DIDResolver } from '../../../did/did-resolver';
+import { verify as verifyEd25519 } from '../../algorithms/ed25519';
+import { verify as verifySecp256k1 } from '../../algorithms/secp256k1';
+
+const verifiers: { [key:string]: VerifyFn } = {
+  'Ed25519'   : verifyEd25519,
+  'secp256k1' : verifySecp256k1
+};
+
+type VerificationResult = {
+  signers: string[];
+};
+
+// TODO: add logic to prevent validating duplicate signatures
+export class GeneralJwsVerifier {
+  jws: GeneralJws;
+
+  constructor(jws: GeneralJws) {
+    this.jws = jws;
+  }
+
+  async verify(didResolver: DIDResolver): Promise<VerificationResult> {
+    const signers: string[] = [];
+
+    for (let signature of this.jws.signatures) {
+      const protectedBytes = base64url.baseDecode(signature.protected);
+      const protectedJson = new TextDecoder().decode(protectedBytes);
+
+      const { kid } = JSON.parse(protectedJson);
+      const did = this.extractDid(kid);
+      const publicKeyJwk = await this.getPublicKey(did, kid, didResolver);
+
+      const isVerified = await this.verifySignature(this.jws.payload, signature, publicKeyJwk);
+
+      if (isVerified) {
+        signers.push(did);
+      } else {
+        throw new Error('signature verification failed');
+      }
+    }
+
+    return { signers };
+  }
+
+  async getPublicKey(did: string, kid: string, didResolver: DIDResolver): Promise<JwkPublic> {
+    // `resolve` throws exception if DID is invalid, DID method is not supported,
+    // or resolving DID fails
+
+    const { didDocument } = await didResolver.resolve(did);
+    const { verificationMethod: verificationMethods = [] } = didDocument || {};
+
+    let verificationMethod: VerificationMethod | undefined;
+
+    for (const vm of verificationMethods) {
+      // consider optimizing using a set for O(1) lookups if needed
+      if (vm.id === kid) {
+        verificationMethod = vm;
+        break;
+      }
+    }
+
+
+    if (!verificationMethod) {
+      throw new Error('public key needed to verify signature not found in DID Document');
+    }
+
+    // TODO: replace with JSON Schema based validation
+    // more info about the `JsonWebKey2020` type can be found here:
+    // https://www.w3.org/TR/did-spec-registries/#jsonwebkey2020
+    if (verificationMethod.type !== 'JsonWebKey2020') {
+      throw new Error(`verification method [${kid}] must be JsonWebKey2020`);
+    }
```

Comment:
JFYI. Looks like one more high(er) priority updaet to ION/ION SDK.... this has been a low-pri in the backlog. The SDK still uses `EcdsaSecp256k1VerificationKey2019` for sekp256k1.... should be a simple one line fix?! Just bumped the [issue](https://github.com/decentralized-identity/ion/issues/229) priority.

Reply:
seems like we'll still want to support `EcdsaSecp256k1VerificationKey2019` though yeah? in order to support DIDs created prior to that issue getting resolved?

Reply:
@csuwildcat

Reply:
we can open an issue for this if we need to add support for it